### PR TITLE
Make m*deleg regs conditional on U/S/N

### PR DESCRIPTION
**Pull request by timsifive**
Change-Id: I544fc15625400d8ad64d4a65f0fc9d77f428ca84

### CODE REVIEW

File: src/target/riscv/riscv.c [hunk 1]
```diff
@@ -2323,6 +2323,15 @@ int riscv_init_registers(struct target *target)
 				case CSR_SATP:
 					r->exist = riscv_supports_extension(target, 'S');
 					break;
+				case CSR_MEDELEG:
```
Comment (mwachs5) [inline src/target/riscv/riscv.c:2326]: this looks good, do you generally also try to read the SEDELEG or SIDELEG registers? The same limitation about 'N' extension applies there.

**Review by mwachs5 — APPROVED** [review]

**Review by mwachs5 — CHANGES_REQUESTED** [review]
actually wait I think the logic is wrong...

File: src/target/riscv/riscv.c [hunk 2]
```diff
@@ -2323,6 +2323,15 @@ int riscv_init_registers(struct target *target)
...
+				case CSR_MIDELEG:
+					/* "In systems with only M-mode, or with both M-mode and
+					 * U-mode but without U-mode trap support, the medeleg and
+					 * mideleg registers should not exist." */
+					r->exist = (riscv_supports_extension(target, 'S') ||
```
Comment (mwachs5) [inline src/target/riscv/riscv.c]: this should be `r->exist = (supports(S) || supports (N))`. If you have `N`, you SHOULD have this register, which is not what the code currently does.

File: src/target/riscv/riscv.c (hunk 1 above)
Comment (mwachs5) [reply src/target/riscv/riscv.c:2326]: I guess they are not there at all, maybe they should be with the caveat they only appear if ` supports(S) && supports(N)`

File: src/target/riscv/riscv.c (hunk 2 above)
Comment (timsifive) [reply src/target/riscv/riscv.c]: Right. I tried to derive this again. Can you check my logic, starting with what it says in the spec (I think):
                    r->exist = !((M && !U && !S) || (M && U && !N))
                             = !(M && !U && !S) && !(M && U && !N)
                             = (!M || U || S) && (!M || !U || N)
                             = (U || S) && (!U || N)

File: src/target/riscv/riscv.c (hunk 1 above)
Comment (timsifive) [reply src/target/riscv/riscv.c:2326]: s*deleg aren't in riscv-opcodes, which is where my list of registers comes from. Should they be?

Comment (mwachs5) [reply src/target/riscv/riscv.c:2326]: I think since nothing implements `N` yet they're probably missing

File: src/target/riscv/riscv.c (hunk 2 above)
Comment (mwachs5) [reply src/target/riscv/riscv.c]: I think the phrase "or with both M-Mode and U-Mode  but without U-mode trap support" has an implied "only"... as in, "only M-Mode and U-Mode, but without U-mode trap support".

Also there are other implied things like you can't have S without U, and you can't have N without U (presumably stated somewhere else in the spec).

**Comment by timsifive** [comment]
Can I get a review for this?

**Review by mwachs5 — APPROVED** [review]
LGTM. I haven't tested it but the logic looks right now!

**Comment by timsifive** [comment]
Thanks, Megan!

**Pull request merged**